### PR TITLE
chore: cloudwatch metrics example

### PR DESCRIPTION
**Pull request by anitarua**
Addresses [#442](https://github.com/momentohq/dev-eco-issue-tracker/issues/442). Created a new `lambda-examples` directory to show how a CDK stack can be configured to deploy a Lambda function or ECS cluster, create metric filters for the associated log groups (or a custom log group), and create a CloudWatch dashboard. 

The dashboard looks something like this when populated (updated dashboard):
cloudwatch dashboard

The dashboard uses values in the logs emitted by both the [Nodejs SDK metrics](https://github.com/momentohq/client-sdk-javascript/blob/main/packages/client-sdk-nodejs/src/config/middleware/impl/experimental-metrics-middleware.ts#L26) and [.NET SDK metrics](https://github.com/momentohq/client-sdk-dotnet/blob/2900922d8bd46ac3f91cb06dbfe070419dc9b738/src/Momento.Sdk/Config/Middleware/ExperimentalMetricsMiddleware.cs#L18) interfaces.

Also tested that this works with a .NET lambda function (that work currently exists in an [unmerged branch](https://github.com/anitarua/client-sdk-dotnet/tree/dotnet-lambda-example) but can merge into the dotnet repo if it seems useful).

Let me know if you think the dashboard should include/exclude any other metrics!

Original dashboard when PR opened:
metrics example dashboard

### CODE REVIEW

File: examples/nodejs/lambda-examples/cloudwatch-metrics/README.md
Comment (bruuuuuuuce) [inline examples/nodejs/lambda-examples/cloudwatch-metrics/README.md:1]: Your readme/docs are always on point, nice job 🤜

File: examples/nodejs/lambda-examples/cloudwatch-metrics/infrastructure/lib/momento-metrics-stack.ts
Comment (bruuuuuuuce) [inline examples/nodejs/lambda-examples/cloudwatch-metrics/infrastructure/lib/momento-metrics-stack.ts:62]: nit: we should be using nodejs 18 or 20, nodejs16 is EOL

File: examples/nodejs/lambda-examples/cloudwatch-metrics/infrastructure/lib/momento-metrics-stack.ts
Comment (bruuuuuuuce) [inline examples/nodejs/lambda-examples/cloudwatch-metrics/infrastructure/lib/momento-metrics-stack.ts:100]: nit: we know the secret name for the momento api key above. We should restrict listing/getting secret values to only that secret

File: examples/nodejs/lambda-examples/cloudwatch-metrics/infrastructure/lib/momento-metrics-stack.ts
Comment (bruuuuuuuce) [inline examples/nodejs/lambda-examples/cloudwatch-metrics/infrastructure/lib/momento-metrics-stack.ts:111]: ditto

File: examples/nodejs/lambda-examples/cloudwatch-metrics/lambda/handler.ts
Comment (bruuuuuuuce) [inline examples/nodejs/lambda-examples/cloudwatch-metrics/lambda/handler.ts:24]: momento api calls should be fast so this should probably not matter, but just calling out that this isnt actually 5 min of gets and sets. Its 300 gets and sets plus a 300 1-second delays. So its 300 x (1 + epsilon) seconds.

File: examples/nodejs/lambda-examples/cloudwatch-metrics/lambda/handler.ts
Comment (bruuuuuuuce) [inline examples/nodejs/lambda-examples/cloudwatch-metrics/lambda/handler.ts:17]: do we log a ton when debug logging? If we do we might want to make this `INFO`. If not then `DEBUG` is fine

Comment (bruuuuuuuce) [reply examples/nodejs/lambda-examples/cloudwatch-metrics/lambda/handler.ts:17]: O is the `DEBUG` for the middleware, so we can grep the logs and make a dashboard

File: examples/nodejs/lambda-examples/cloudwatch-metrics/infrastructure/bin/infrastructure.ts
Comment (bruuuuuuuce) [inline examples/nodejs/lambda-examples/cloudwatch-metrics/infrastructure/bin/infrastructure.ts:15]: nit: looks like we are hardcoding the region here to `us-west-2`, so even if a user passes in a different env, we will still have resources in west-2. I don't think it would be hard to change this, or we can just remove this text and say it will only deploy to `us-west-2`

File: examples/nodejs/lambda-examples/cloudwatch-metrics/infrastructure/lib/momento-metrics-stack.ts
Comment (bruuuuuuuce) [inline examples/nodejs/lambda-examples/cloudwatch-metrics/infrastructure/lib/momento-metrics-stack.ts:129]: This is cool I didn't know you could do this

File: examples/nodejs/lambda-examples/cloudwatch-metrics/docker/ecs-code/index.ts
Comment (cprice404) [inline examples/nodejs/lambda-examples/cloudwatch-metrics/docker/ecs-code/index.ts:44]: just a nit but curious why this eslint hint is necessary here?

File: examples/nodejs/lambda-examples/cloudwatch-metrics/docker/Dockerfile
Comment (cprice404) [inline examples/nodejs/lambda-examples/cloudwatch-metrics/docker/Dockerfile:10]: what happens if you leave this out?

File: examples/nodejs/lambda-examples/cloudwatch-metrics/infrastructure/bin/infrastructure.ts
Comment (cprice404) [reply examples/nodejs/lambda-examples/cloudwatch-metrics/infrastructure/bin/infrastructure.ts:15]: good catch @bruuuuuuuce . let's see if we can get rid of the hard-coded us-west-2 refs

File: examples/nodejs/lambda-examples/cloudwatch-metrics/infrastructure/lib/momento-metrics-stack.ts
Comment (cprice404) [inline examples/nodejs/lambda-examples/cloudwatch-metrics/infrastructure/lib/momento-metrics-stack.ts:58]: I would love to move all of the stuff that has to do with deploying the lambda/ecs examples to a few smaller, purpose-built functions so that they don't distract from the flow of readability for this main code.

File: examples/nodejs/lambda-examples/cloudwatch-metrics/infrastructure/lib/momento-metrics-stack.ts
Comment (cprice404) [inline examples/nodejs/lambda-examples/cloudwatch-metrics/infrastructure/lib/momento-metrics-stack.ts:129]: would probably be useful to add a code comment above this part, explaining what it is doing. "The Momento experimental metrics middleware produces JSON logs with metrics about each Momento request.  An example log entry looks like this: ...

In this section we use CloudWatch Metric Filters (maybe link to docs) to extract metrics from these log messages."

File: examples/nodejs/lambda-examples/cloudwatch-metrics/infrastructure/lib/momento-metrics-stack.ts
Comment (cprice404) [inline examples/nodejs/lambda-examples/cloudwatch-metrics/infrastructure/lib/momento-metrics-stack.ts:170]: the `cdk.Stack` object should have a variable that you can read to get the region string. I can't remember if it's `this.region` or `this.env.region` or what, but it should be available.

File: examples/nodejs/lambda-examples/cloudwatch-metrics/infrastructure/lib/momento-metrics-stack.ts
Comment (cprice404) [inline examples/nodejs/lambda-examples/cloudwatch-metrics/infrastructure/lib/momento-metrics-stack.ts:191]: I suspect that some of the layout properties you are setting on some of these widgets may align with the defaults.

In the past I have found that it's easier to manage programmatic dashboards if you keep the number of layout properties that you explicitly specify to a minimum. You will definitely need some height/width/position attributes, and you may also want to keep some of your annotation stuff. Heck, maybe you need everything here :) but just wanted to call out - i would experiment with omitting anything you don't think is absolutely necessary and see if it still looks okay, and if so, omit what you can.

totally your call on which things clear the bar of being necessary though!

**Review by cprice404 — COMMENTED** [review]
this is looking awesome!

i want to test it in my account before we merge it, and not sure if i will get to that today. but i left a handful of minor feedback that I think you could go ahead and move forward on when you have bandwidth.

File: examples/nodejs/lambda-examples/cloudwatch-metrics/README.md
Comment (anitarua) [reply examples/nodejs/lambda-examples/cloudwatch-metrics/README.md]: Couldn't get the `stackConfig` variable working with CFN stack parameters (they're lazily-loaded tokens but the constructor needs to know the value while creating the stack). I instead moved all the config related stuff to a `.env` file and updated the README with instructions for those.

File: examples/nodejs/lambda-examples/cloudwatch-metrics/docker/ecs-code/index.ts
Comment (anitarua) [reply examples/nodejs/lambda-examples/cloudwatch-metrics/docker/ecs-code/index.ts:44]: can't recall why anymore, but I've removed the hints

File: examples/nodejs/lambda-examples/cloudwatch-metrics/README.md
Comment (cprice404) [inline examples/nodejs/lambda-examples/cloudwatch-metrics/README.md:21]:
```suggestion
## Deploying the CloudWatch Dashboard (along with an example Momento Lambda Function)
```

File: examples/nodejs/lambda-examples/cloudwatch-metrics/README.md
Comment (cprice404) [inline examples/nodejs/lambda-examples/cloudwatch-metrics/README.md:45]: nit - can we rename this var to something like `EXAMPLE_MOMENTO_APPLICATION` or `ADDITIONAL_APPLICATION` or something along those lines?

**Review by cprice404 — COMMENTED** [review]
anita i deployed this locally and started playing with it, but ran out of time. I will finish reviewing this evening!

File: examples/nodejs/lambda-examples/cloudwatch-metrics/docker/ecs-code/index.ts
Comment (cprice404) [reply examples/nodejs/lambda-examples/cloudwatch-metrics/docker/ecs-code/index.ts:24]: actually ended up with this:

```
const delayBetweenRequestsMillis = 100;
    const delaysPerSecond = 1000 / delayBetweenRequestsMillis;
    for (let i = 0; i < 60 /* seconds */ * 60 /* minutes */ * delaysPerSecond; i++) {
      await cacheClient.set('cache', 'metrics-example-{i}', 'VALUE');
      await cacheClient.get('cache', 'metrics-example-{i}');
      await delay(delayBetweenRequestsMillis);
    }
```

File: examples/nodejs/lambda-examples/cloudwatch-metrics/infrastructure/lib/momento-metrics-stack.ts
Comment (cprice404) [inline examples/nodejs/lambda-examples/cloudwatch-metrics/infrastructure/lib/momento-metrics-stack.ts:95]: woof, is this what really shows up for the request type now? that seems like a bug in my last refactor. We want this field to show something like `Get`, `Set`, `ListPushBack`, etc. If all of the requests are showing up with just this, then we should make a quick tweak on `main` to fix that and then rebase your branch. I can help with that if you need another pair of eyes.

File: examples/nodejs/lambda-examples/cloudwatch-metrics/infrastructure/lib/momento-metrics-stack.ts
Comment (cprice404) [inline examples/nodejs/lambda-examples/cloudwatch-metrics/infrastructure/lib/momento-metrics-stack.ts:192]: see my comments below about `dimensions` before reading this.

For this metric, adding a dimension for the request type would be extremely useful. Then we could graph p50/p99 request size / response size by request type, for a few of the most common request types.

This also would rely on fixing the bug I mentioned above about how the request type is currently showing up as 'middleware' or whatever.

File: examples/nodejs/lambda-examples/cloudwatch-metrics/infrastructure/lib/momento-metrics-stack.ts
Comment (cprice404) [inline examples/nodejs/lambda-examples/cloudwatch-metrics/infrastructure/lib/momento-metrics-stack.ts:274]: Let's call this "requests per second".

File: examples/nodejs/lambda-examples/cloudwatch-metrics/infrastructure/lib/momento-metrics-stack.ts
Comment (cprice404) [inline examples/nodejs/lambda-examples/cloudwatch-metrics/infrastructure/lib/momento-metrics-stack.ts:277]: i'm curious what is the intended difference between this widget and the previous? They are both basically graphing requests per second and I'm seeing the same values on the two graphs in my dashboard.

I also don't completely understand why the values are showing up as the same, when on one you are graphing sample count over the last minute and one you are graphing sample count over the last second :) something fishy about that that i don't know the answer to. So we need to sort that out, but probably also only need one of the two of these.

Also... I'm sleeping 100ms between requests, so I would expect to see about 10tps * 2 (one get and one set) = 20 tps, but I'm seeing about 1.1k, which seems way too high, and I don't have an answer for that yet either :)

Comment (cprice404) [reply examples/nodejs/lambda-examples/cloudwatch-metrics/infrastructure/lib/momento-metrics-stack.ts:277]: when i look at the lightstep metrics for this I see about 9.5 tps for get and 9.5 tps for set, which is much closer to what i was expecting: https://app.lightstep.com/momento-preprod/dashboard/user-experience/TJXd7Dq4?time_window=minutes_60&chart_id=XYbgq5zQ

Comment (cprice404) [reply examples/nodejs/lambda-examples/cloudwatch-metrics/infrastructure/lib/momento-metrics-stack.ts:277]: I think maybe what is happening here is that the maximum resolution for metricfilter metrics is 1 minute, so even when you set the period to something < 1 minute, you still get the resolution of 1 minute. (we should google this and see if this is documented for cloudwatch metric filters.)

if that's the case, then setting the period to 1m and dividing the value by 60 via cloudwatch metricmath may be the way to go. 1100 / 60 = 18.3, which is approximately what I'm seeing in lightstep for get + set combined. (with my own sleep between requests set to 100ms.)

File: examples/nodejs/lambda-examples/cloudwatch-metrics/infrastructure/lib/momento-metrics-stack.ts
Comment (cprice404) [inline examples/nodejs/lambda-examples/cloudwatch-metrics/infrastructure/lib/momento-metrics-stack.ts:291]: let's do p50 and p99 on this one. and if we are able to get the request type dimension added then I think we should do "get p50", "get p99", "set p50", "set p99".

**Review by cprice404 — COMMENTED** [review]
this is looking really awesome. I left some comments for us to fine tune the dashboard widgets in this last mile, to make them more realistic for an operator dashboard, but this is really great and I can see the finish line!

File: examples/nodejs/lambda-examples/cloudwatch-metrics/infrastructure/lib/momento-metrics-stack.ts
Comment (anitarua) [reply examples/nodejs/lambda-examples/cloudwatch-metrics/infrastructure/lib/momento-metrics-stack.ts:95]: Good catch! Got a PR for this: #993

**Comment by anitarua** [comment]
Lots of great feedback, thanks!! Here's what the dashboard looks like now

cloudwatch dashboard

**Review by cprice404 — APPROVED** [review]
dance

**Pull request merged**
